# STEP3: Perform some simple data analysis

Start by connecting to the database by running the cells below. If you are coming back to this exercise, then uncomment and run the first cell to recreate the database. If you recently completed steps 1 and 2, then skip to the second cell.

In [4]:
!PGPASSWORD=student createdb -h 127.0.0.1 -U student pagila
!PGPASSWORD=student psql -q -h 127.0.0.1 -U student -d pagila -f Data/pagila-schema.sql
!PGPASSWORD=student psql -q -h 127.0.0.1 -U student -d pagila -f Data/pagila-data.sql

 setval 
--------
    200
(1 row)

 setval 
--------
    605
(1 row)

 setval 
--------
     16
(1 row)

 setval 
--------
    600
(1 row)

 setval 
--------
    109
(1 row)

 setval 
--------
    599
(1 row)

 setval 
--------
   1000
(1 row)

 setval 
--------
   4581
(1 row)

 setval 
--------
      6
(1 row)

 setval 
--------
  32098
(1 row)

 setval 
--------
  16049
(1 row)

 setval 
--------
      2
(1 row)

 setval 
--------
      2
(1 row)



In [5]:
%load_ext sql

DB_ENDPOINT = "127.0.0.1"
DB = 'pagila'
DB_USER = 'student'
DB_PASSWORD = 'student'
DB_PORT = '5432'

# postgresql://username:password@host:port/database
conn_string = "postgresql://{}:{}@{}:{}/{}" \
                        .format(DB_USER, DB_PASSWORD, DB_ENDPOINT, DB_PORT, DB)

print(conn_string)
%sql $conn_string

The sql extension is already loaded. To reload it, use:
  %reload_ext sql
postgresql://student:student@127.0.0.1:5432/pagila


'Connected: student@pagila'

### 3NF - Entity Relationship Diagram

<img src="./pagila-3nf.png" width="50%"/>

## 3.1 Insight 1:   Top Grossing Movies 
- Payments amounts are in table `payment`
- Movies are in table `film`
- They are not directly linked, `payment` refers to a `rental`, `rental` refers to an `inventory` item and `inventory` item refers to a `film`
- `payment` &rarr; `rental` &rarr; `inventory` &rarr; `film`

### 3.1.1 Films

In [6]:
%%sql
select film_id, title, release_year, rental_rate, rating  from film limit 5;


 * postgresql://student:***@127.0.0.1:5432/pagila
5 rows affected.


[(1, 'ACADEMY DINOSAUR', 2006, Decimal('0.99'), 'PG'),
 (2, 'ACE GOLDFINGER', 2006, Decimal('4.99'), 'G'),
 (3, 'ADAPTATION HOLES', 2006, Decimal('2.99'), 'NC-17'),
 (4, 'AFFAIR PREJUDICE', 2006, Decimal('2.99'), 'G'),
 (5, 'AFRICAN EGG', 2006, Decimal('2.99'), 'G')]

### 3.1.2 Payments

In [7]:
%%sql
select * from payment limit 5;

 * postgresql://student:***@127.0.0.1:5432/pagila
5 rows affected.


[(16050, 269, 2, 7, Decimal('1.99'), datetime.datetime(2017, 1, 24, 21, 40, 19, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None))),
 (16051, 269, 1, 98, Decimal('0.99'), datetime.datetime(2017, 1, 25, 15, 16, 50, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None))),
 (16052, 269, 2, 678, Decimal('6.99'), datetime.datetime(2017, 1, 28, 21, 44, 14, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None))),
 (16053, 269, 2, 703, Decimal('0.99'), datetime.datetime(2017, 1, 29, 0, 58, 2, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None))),
 (16054, 269, 1, 750, Decimal('4.99'), datetime.datetime(2017, 1, 29, 8, 10, 6, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)))]

### 3.1.3 Inventory

In [8]:
%%sql
select * from inventory limit 5;

 * postgresql://student:***@127.0.0.1:5432/pagila
5 rows affected.


[(1, 1, 1, datetime.datetime(2017, 2, 15, 10, 9, 17, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None))),
 (2, 1, 1, datetime.datetime(2017, 2, 15, 10, 9, 17, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None))),
 (3, 1, 1, datetime.datetime(2017, 2, 15, 10, 9, 17, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None))),
 (4, 1, 1, datetime.datetime(2017, 2, 15, 10, 9, 17, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None))),
 (5, 1, 2, datetime.datetime(2017, 2, 15, 10, 9, 17, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)))]

### 3.1.4 Get the movie of every payment

In [ ]:
%%sql
SELECT f.title, p.amount, p.payment_date, p.customer_id                                            
FROM payment p
JOIN rental r    ON ( p.rental_id = r.rental_id )
JOIN inventory i ON ( r.inventory_id = i.inventory_id )
JOIN film f ON ( i.film_id = f.film_id)
limit 5;

### 3.1.5 sum movie rental revenue
TODO: Write a query that displays the amount of revenue from each title. Limit the results to the top 10 grossing titles. Your results should match the table below.

In [12]:
%%sql
SELECT f.title, SUM(p.amount) as revenue
FROM payment p
JOIN rental r    ON ( p.rental_id = r.rental_id )
JOIN inventory i ON ( r.inventory_id = i.inventory_id )
JOIN film f ON ( i.film_id = f.film_id)
GROUP BY f.title
ORDER BY revenue DESC
limit 10;

 * postgresql://student:***@127.0.0.1:5432/pagila
10 rows affected.


[('TELEGRAPH VOYAGE', Decimal('231.73')),
 ('WIFE TURN', Decimal('223.69')),
 ('ZORRO ARK', Decimal('214.69')),
 ('GOODFELLAS SALUTE', Decimal('209.69')),
 ('SATURDAY LAMBS', Decimal('204.72')),
 ('TITANS JERK', Decimal('201.71')),
 ('TORQUE BOUND', Decimal('198.72')),
 ('HARRY IDAHO', Decimal('195.70')),
 ('INNOCENT USUAL', Decimal('191.74')),
 ('HUSTLER PARTY', Decimal('190.78'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>title</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>TELEGRAPH VOYAGE</td>
        <td>231.73</td>
    </tr>
    <tr>
        <td>WIFE TURN</td>
        <td>223.69</td>
    </tr>
    <tr>
        <td>ZORRO ARK</td>
        <td>214.69</td>
    </tr>
    <tr>
        <td>GOODFELLAS SALUTE</td>
        <td>209.69</td>
    </tr>
    <tr>
        <td>SATURDAY LAMBS</td>
        <td>204.72</td>
    </tr>
    <tr>
        <td>TITANS JERK</td>
        <td>201.71</td>
    </tr>
    <tr>
        <td>TORQUE BOUND</td>
        <td>198.72</td>
    </tr>
    <tr>
        <td>HARRY IDAHO</td>
        <td>195.70</td>
    </tr>
    <tr>
        <td>INNOCENT USUAL</td>
        <td>191.74</td>
    </tr>
    <tr>
        <td>HUSTLER PARTY</td>
        <td>190.78</td>
    </tr>
</tbody></table></div>

## 3.2 Insight 2:   Top grossing cities 
- Payments amounts are in table `payment`
- Cities are in table `cities`
- `payment` &rarr; `customer` &rarr; `address` &rarr; `city`

### 3.2.1 Get the city of each payment

In [21]:
%%sql
SELECT ci.city, p.amount as revenue
FROM payment p
JOIN customer c  ON ( p.customer_id = c.customer_id )
JOIN address a ON ( c.address_id = a.address_id )
JOIN city ci ON ( a.city_id = ci.city_id )
GROUP BY ci.city, p.amount

 * postgresql://student:***@127.0.0.1:5432/pagila
4804 rows affected.


[('Basel', Decimal('0.99')),
 ('Cuman', Decimal('7.99')),
 ('Erlangen', Decimal('5.99')),
 ('Iwaki', Decimal('2.99')),
 ('Kolpino', Decimal('3.99')),
 ('Pangkal Pinang', Decimal('2.99')),
 ('Tama', Decimal('7.99')),
 ('Kolpino', Decimal('6.99')),
 ('Bandar Seri Begawan', Decimal('5.99')),
 ('Bag', Decimal('2.99')),
 ('Kimchon', Decimal('0.99')),
 ('Jinchang', Decimal('10.99')),
 ('Cavite', Decimal('0.99')),
 ('Ziguinchor', Decimal('7.99')),
 ('Citrus Heights', Decimal('0.99')),
 ('Tychy', Decimal('1.99')),
 ('Jhansi', Decimal('4.99')),
 ('Maikop', Decimal('0.99')),
 ('Nha Trang', Decimal('7.99')),
 ('Saarbrcken', Decimal('4.99')),
 ('Hagonoy', Decimal('8.99')),
 ('al-Qatif', Decimal('4.99')),
 ('Nyeri', Decimal('8.99')),
 ('Patras', Decimal('2.99')),
 ('Nagareyama', Decimal('9.99')),
 ('Deba Habe', Decimal('7.99')),
 ('Shikarpur', Decimal('5.99')),
 ('Braslia', Decimal('4.99')),
 ('Ezeiza', Decimal('3.99')),
 ('Bucuresti', Decimal('1.99')),
 ('Cam Ranh', Decimal('5.99')),
 ('Fengshan', Decimal('1.99')),
 ('Abha', Decimal('0.99')),
 ('Ljubertsy', Decimal('6.99')),
 ('Alessandria', Decimal('5.99')),
 ('Ljubertsy', Decimal('3.99')),
 ('Augusta-Richmond County', Decimal('0.99')),
 ('Konotop', Decimal('0.99')),
 ('Tongliao', Decimal('4.99')),
 ('Berhampore (Baharampur)', Decimal('5.99')),
 ('Koriyama', Decimal('1.99')),
 ('Ranchi', Decimal('0.99')),
 ('Omiya', Decimal('10.99')),
 ('Huaian', Decimal('3.99')),
 ('Allende', Decimal('2.99')),
 ('Linz', Decimal('0.99')),
 ('Tarsus', Decimal('2.99')),
 ('al-Hawiya', Decimal('9.99')),
 ('Boksburg', Decimal('4.99')),
 ('Tegal', Decimal('1.99')),
 ('Rockford', Decimal('10.99')),
 ('Mwanza', Decimal('3.99')),
 ('Mwanza', Decimal('6.99')),
 ('So Leopoldo', Decimal('8.99')),
 ('Vilnius', Decimal('0.99')),
 ('Aurora', Decimal('0.99')),
 ('Zalantun', Decimal('4.99')),
 ('Bergamo', Decimal('5.99')),
 ('Kurashiki', Decimal('0.99')),
 ('Nam Dinh', Decimal('3.99')),
 ('Suihua', Decimal('4.99')),
 ('Tanshui', Decimal('6.99')),
 ('Bellevue', Decimal('4.99')),
 ('Tanshui', Decimal('3.99')),
 ('Nam Dinh', Decimal('6.99')),
 ('Bradford', Decimal('5.99')),
 ('Laohekou', Decimal('4.99')),
 ('Kamyin', Decimal('1.99')),
 ('Chungho', Decimal('8.99')),
 ('Arecibo', Decimal('5.99')),
 ('Jurez', Decimal('2.99')),
 ('Uijongbu', Decimal('0.99')),
 ('Karnal', Decimal('7.99')),
 ('Kursk', Decimal('8.99')),
 ('Fuyu', Decimal('5.99')),
 ('Kuwana', Decimal('8.99')),
 ('Livorno', Decimal('4.99')),
 ('Barcelona', Decimal('6.99')),
 ('Lapu-Lapu', Decimal('0.99')),
 ('Maring', Decimal('2.99')),
 ('Ivanovo', Decimal('2.99')),
 ('Ipoh', Decimal('6.99')),
 ('Fukuyama', Decimal('5.99')),
 ('Ipoh', Decimal('3.99')),
 ('Namibe', Decimal('0.99')),
 ('Salala', Decimal('6.99')),
 ('Santa F', Decimal('2.99')),
 ('Sullana', Decimal('2.99')),
 ('Ambattur', Decimal('7.99')),
 ('Elista', Decimal('1.99')),
 ('Guadalajara', Decimal('5.99')),
 ('Baybay', Decimal('4.99')),
 ('Roanoke', Decimal('0.99')),
 ('Benin City', Decimal('4.99')),
 ('Battambang', Decimal('5.99')),
 ('Rustenburg', Decimal('0.99')),
 ('Halisahar', Decimal('3.99')),
 ('Pune', Decimal('4.99')),
 ('Nagaon', Decimal('5.99')),
 ('Citt del Vaticano', Decimal('9.99')),
 ('Adana', Decimal('0.99')),
 ('Halisahar', Decimal('6.99')),
 ('London', Decimal('5.99')),
 ('Sirjan', Decimal('8.99')),
 ('Taizz', Decimal('2.99')),
 ('Angra dos Reis', Decimal('7.99')),
 ('Ocumare del Tuy', Decimal('8.99')),
 ('Tiefa', Decimal('4.99')),
 ('Tanauan', Decimal('8.99')),
 ('Chatsworth', Decimal('2.99')),
 ('Hanoi', Decimal('4.99')),
 ('Haldia', Decimal('4.99')),
 ('Ibirit', Decimal('6.99')),
 ('Iwatsuki', Decimal('0.99')),
 ('Ibirit', Decimal('3.99')),
 ('Sawhaj', Decimal('3.99')),
 ('Ashdod', Decimal('1.99')),
 ('Sawhaj', Decimal('6.99')),
 ('Qomsheh', Decimal('4.99')),
 ('Bamenda', Decimal('0.99')),
 ('Vaduz', Decimal('4.99')),
 ('Athenai', Decimal('2.99')),
 ('Amroha', Decimal('1.99')),
 ('Gaziantep', Decimal('2.99')),
 ('Kamakura', Decimal('2.99')),
 ('Smolensk', Decimal('6.99')),
 ('Malm', D

### 3.2.2 Top grossing cities
TODO: Write a query that returns the total amount of revenue by city as measured by the `amount` variable in the `payment` table. Limit the results to the top 10 cities. Your result should match the table below.

In [25]:
%%sql
SELECT ci.city, SUM(p.amount) as revenue
FROM payment p
JOIN customer c  ON ( p.customer_id = c.customer_id )
JOIN address a ON ( c.address_id = a.address_id )
JOIN city ci ON ( a.city_id = ci.city_id )
GROUP BY ci.city
ORDER BY revenue DESC
LIMIT 10;

 * postgresql://student:***@127.0.0.1:5432/pagila
10 rows affected.


[('Cape Coral', Decimal('221.55')),
 ('Saint-Denis', Decimal('216.54')),
 ('Aurora', Decimal('198.50')),
 ('Molodetno', Decimal('195.58')),
 ('Apeldoorn', Decimal('194.61')),
 ('Santa Brbara dOeste', Decimal('194.61')),
 ('Qomsheh', Decimal('186.62')),
 ('London', Decimal('180.52')),
 ('Ourense (Orense)', Decimal('177.60')),
 ('Bijapur', Decimal('175.61'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>city</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>Cape Coral</td>
        <td>221.55</td>
    </tr>
    <tr>
        <td>Saint-Denis</td>
        <td>216.54</td>
    </tr>
    <tr>
        <td>Aurora</td>
        <td>198.50</td>
    </tr>
    <tr>
        <td>Molodetno</td>
        <td>195.58</td>
    </tr>
    <tr>
        <td>Apeldoorn</td>
        <td>194.61</td>
    </tr>
    <tr>
        <td>Santa Brbara dOeste</td>
        <td>194.61</td>
    </tr>
    <tr>
        <td>Qomsheh</td>
        <td>186.62</td>
    </tr>
    <tr>
        <td>London</td>
        <td>180.52</td>
    </tr>
    <tr>
        <td>Ourense (Orense)</td>
        <td>177.60</td>
    </tr>
    <tr>
        <td>Bijapur</td>
        <td>175.61</td>
    </tr>
</tbody></table></div>

## 3.3 Insight 3 : Revenue of a movie by customer city and by month 

### 3.3.1 Total revenue by month

In [ ]:
%%sql
SELECT sum(p.amount) as revenue, EXTRACT(month FROM p.payment_date) as month
from payment p
group by month
order by revenue desc
limit 10;

### 3.3.2 Each movie by customer city and by month (data cube)

In [ ]:
%%sql
SELECT f.title, p.amount, p.customer_id, ci.city, p.payment_date,EXTRACT(month FROM p.payment_date) as month
FROM payment p
JOIN rental r    ON ( p.rental_id = r.rental_id )
JOIN inventory i ON ( r.inventory_id = i.inventory_id )
JOIN film f ON ( i.film_id = f.film_id)
JOIN customer c  ON ( p.customer_id = c.customer_id )
JOIN address a ON ( c.address_id = a.address_id )
JOIN city ci ON ( a.city_id = ci.city_id )
order by p.payment_date
limit 10;

### 3.3.3 Sum of revenue of each movie by customer city and by month

TODO: Write a query that returns the total amount of revenue for each movie by customer city and by month. Limit the results to the top 10 movies. Your result should match the table below.

In [103]:
%%sql
SELECT f.title,  ci.city, EXTRACT(month FROM p.payment_date) as month, SUM(p.amount) as revenue
FROM payment p
JOIN rental r  ON ( p.rental_id = r.rental_id )
JOIN inventory i ON ( r.inventory_id = i.inventory_id )
JOIN film f ON ( i.film_id = f.film_id)
JOIN customer c  ON ( p.customer_id = c.customer_id )
JOIN address a ON ( c.address_id = a.address_id )
JOIN city ci ON ( a.city_id = ci.city_id )
GROUP BY city, month, title
ORDER BY month ASC,  revenue DESC, TITLE, city
LIMIT 10;

 * postgresql://student:***@127.0.0.1:5432/pagila
10 rows affected.


[('SHOW LORD', 'Mannheim', 1.0, Decimal('11.99')),
 ('AMERICAN CIRCUS', 'Callao', 1.0, Decimal('10.99')),
 ('CASUALTIES ENCINO', 'Warren', 1.0, Decimal('10.99')),
 ('KISSING DOLLS', 'Toulon', 1.0, Decimal('10.99')),
 ('TELEGRAPH VOYAGE', 'Naala-Porto', 1.0, Decimal('10.99')),
 ('AUTUMN CROW', 'Ashgabat', 1.0, Decimal('9.99')),
 ('CALIFORNIA BIRDS', 'Plock', 1.0, Decimal('9.99')),
 ('DARKO DORADO', 'Bhilwara', 1.0, Decimal('9.99')),
 ('DAY UNFAITHFUL', 'Baybay', 1.0, Decimal('9.99')),
 ('HEAD STRANGER', 'Xiangtan', 1.0, Decimal('9.99'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>title</th>
        <th>city</th>
        <th>month</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>SHOW LORD</td>
        <td>Mannheim</td>
        <td>1.0</td>
        <td>11.99</td>
    </tr>
    <tr>
        <td>AMERICAN CIRCUS</td>
        <td>Callao</td>
        <td>1.0</td>
        <td>10.99</td>
    </tr>
    <tr>
        <td>CASUALTIES ENCINO</td>
        <td>Warren</td>
        <td>1.0</td>
        <td>10.99</td>
    </tr>
    <tr>
        <td>TELEGRAPH VOYAGE</td>
        <td>Naala-Porto</td>
        <td>1.0</td>
        <td>10.99</td>
    </tr>
    <tr>
        <td>KISSING DOLLS</td>
        <td>Toulon</td>
        <td>1.0</td>
        <td>10.99</td>
    </tr>
    <tr>
        <td>MILLION ACE</td>
        <td>Bergamo</td>
        <td>1.0</td>
        <td>9.99</td>
    </tr>
    <tr>
        <td>TITANS JERK</td>
        <td>Kimberley</td>
        <td>1.0</td>
        <td>9.99</td>
    </tr>
    <tr>
        <td>DARKO DORADO</td>
        <td>Bhilwara</td>
        <td>1.0</td>
        <td>9.99</td>
    </tr>
    <tr>
        <td>SUNRISE LEAGUE</td>
        <td>Nagareyama</td>
        <td>1.0</td>
        <td>9.99</td>
    </tr>
    <tr>
        <td>MILLION ACE</td>
        <td>Gaziantep</td>
        <td>1.0</td>
        <td>9.99</td>
    </tr>
</tbody></table></div>